Source/WebCore/Modules/permissions/PermissionStatus.cpp

```diff
@@ -101,4 +150,32 @@ void PermissionStatus::eventListenersDidChange()
     m_hasChangeEventListener = hasEventListeners(eventNames().changeEvent);
 }
 
+MainThreadPermissionObserver::MainThreadPermissionObserver(WeakPtr<PermissionObserver> workerPermissionObserver, ScriptExecutionContextIdentifier workerContextIdentifier, const PermissionDescriptor& workerDescriptor, const ClientOrigin& workerOrigin, PermissionQuerySource source, Page* page, PermissionState state)
+    : m_identifier(nextMainThreadPermissionObserverIdentifier())
+    , m_workerPermissionObserver(workerPermissionObserver)
+    , m_workerContextIdentifier(workerContextIdentifier)
+    , m_workerDescriptor(workerDescriptor)
+    , m_workerOrigin(workerOrigin)
+    , m_source(source)
+    , m_page(page)
+    , m_state(state)
+{
+    PermissionController::shared().addObserver(*this);
+}
+
+MainThreadPermissionObserver::~MainThreadPermissionObserver()
+{
+    PermissionController::shared().removeObserver(*this);
+}
+
+void MainThreadPermissionObserver::stateChanged(PermissionState permissionState) 
+{
+    m_state = permissionState;
+
+    ScriptExecutionContext::postTaskTo(m_workerContextIdentifier, [this, &permissionState](auto&) {
```

Comment:
What guarantees that `this` is still alive when the lambda runs? We should probably just capture `workerPermissionObserver = m_workerPermissionObserver`, not `this`

Reply:
Also, again, you can never capture variables such as `permissionState` by reference in an async lambda.

Reply:
I believe that there is no guarantee that `this` is alive when the lambda runs. `this` is held in the NeverDestroyed map as long as JavaScript keeps the associated `PermissionStatus` object alive. So yeah, it's probably best to not capture `this`. 

But `m_workerPermissionObserver` is a WeakPtr. It's not safe to capture a WeakPtr in a lambda that will be run on another thread right?

Reply:
You can capture and pass WeakPtr to any thread. What you need to be careful about is on what thread you check if the WeakPtr is null or not (Needs to the be same same the object gets destroyed on).

Reply:
Right, but in order to call PermissionObserver::stateChanged in the callback, we need to check if m_workerPermissionObserver is null or not--and this callback runs on the worker thread but the m_workerPermissionObserver WeakPtr was constructed on the main thread.

Reply:
Ah, ok one solution is that I make PermissionObserver threadSafeRefCounted, then instead of having to check if the WeakPtr m_workerPermissionObserver is null in the callback, we can check it on the main thread, if not null, then we can Ref it, pass the Ref to the call back, then re-dispatch the ref back to main thread to be destroyed. Then, instead of storing a uniquePtr to the MainThreadPermissionObserver in the map, we can just store RefPtr's. How's that sound?

Reply:
You are confused. m_workerPermissionObserver refers to a worker object, thus null-checking it and using it on the worker thread is the correct usage. I wouldn't have recommended this if it wasn't correct.